sklearn/preprocessing/_target_encoder.py

```diff
@@ -0,0 +1,165 @@
+            Transformed input.
+        """
+        check_is_fitted(self)
+        X_int, X_known = self._transform(X, handle_unknown='ignore')
+        return self._transform_X_int(X_int, X_known)
+
+    def fit_transform(self, X, y):
+        """Fit the encoder and encodes `X`.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+            The data to encode.
+
+        y : array-like of shape (n_samples,)
+            Target values.
+
+        Returns
+        -------
+        X_out : ndarray of shape (n_samples, n_features)
+            Transformed input.
+        """
+        X_int, X_known = self._fit(X, y)
+        return self._transform_X_int(X_int, X_known)
+
+    def _transform_X_int(self, X_int, X_known):
+        """Transform integer encoded X. `X_int` and `X_known` are returned
+        by `_transform`.
+        """
+        out = np.empty_like(X_int, dtype=float)
+        for i, cat_encoding in enumerate(self.encodings_):
+            np.take(cat_encoding, X_int[:, i], out=out[:, i])
+            out[~X_known[:, i], i] = self.encoding_mean_
+        return out
+
+    def _fit(self, X, y):
+        """Fit the encoder"""
+        y = column_or_1d(y, warn=True)
+        fit_results = super()._fit(X, handle_unknown='ignore',
+                                   return_counts=True)
+        X_int, X_known = self._transform(X, handle_unknown='ignore')
+        # Makes sure unknown categories are not used fot fitting
+        X_int[~X_known] = -1
+        self.encoding_mean_ = y_mean = np.mean(y)
+
+        # y is constant the encoding will be the constant
+        if np.ptp(y) == 0.0:
+            self.encodings_ = [
+                np.full(len(cat), fill_value=y_mean, dtype=float)
+                for cat in self.categories_]
+            return X_int, X_known
+
+        y_variance = np.var(y)
+
+        n_samples, n_features = X_int.shape
+        # temporary location to store masks when encoding each category
+        tmp_mask = np.zeros(n_samples, dtype=bool)
+
+        cat_encodings = []
+        category_counts = fit_results['category_counts']
+
+        for i in range(n_features):
+            n_cats = len(self.categories_[i])
+            cat_means = np.zeros(n_cats, dtype=float)
+            cat_var_ratio = np.ones(n_cats, dtype=float)
+
+            for encoding in range(n_cats):
+                np.equal(X_int[:, i], encoding, out=tmp_mask)
+                y_tmp = y[tmp_mask]
+                if y_tmp.size:
+                    cat_means[encoding] = np.mean(y_tmp)
+                    cat_var_ratio[encoding] = np.var(y_tmp)
+
+            # partial-pooling estimates
+            cat_counts = category_counts[i]
+            cat_var_ratio /= y_variance
+
+            cat_encoded = cat_counts * cat_means + cat_var_ratio * y_mean
+            cat_encoded /= cat_counts + cat_var_ratio
+            cat_encodings.append(cat_encoded)
```

Comment:
I am not sure that this matches the James Stein estimator as described in this post:

https://towardsdatascience.com/benchmarking-categorical-encoders-9c322bd77ee8

Reply:
We need to check how this implementation is related to https://github.com/scikit-learn-contrib/category_encoders/blob/master/category_encoders/james_stein.py